packages/sdk/src/engine-commands/getEnginesMetaInfo.ts

```diff
@@ -0,0 +1,177 @@
+  // if the engine is not found, or the version cannot be retrieved, keep track of the resulting errors.
+  const errors = [] as Error[]
+
+  const resolved = match(enginesInfo)
+    .with({ fromEnvVar: P.when(O.isSome) }, (enginesInfoLibrary) => {
+      return `, resolved by ${enginesInfoLibrary.fromEnvVar.value}`
+    })
+    .otherwise(() => '')
+
+  const version = match(enginesInfo)
+    .with({ version: P.when(E.isRight) }, (engineInfo) => {
+      return engineInfo.version.right
+    })
+    .with({ version: P.when(E.isLeft) }, (engineInfo) => {
+      errors.push(engineInfo.version.left)
+      return 'E_CANNOT_RESOLVE_VERSION' as const
+    })
+    .otherwise(() => {
+      return 'E_CANNOT_RESOLVE_VERSION' as const
+    })
+
+  const absolutePath = match(enginesInfo)
+    .with({ libraryPath: P.when(O.isSome) }, (enginesInfoLibrary) => {
+      return enginesInfoLibrary.libraryPath.value
+    })
+    .with({ libraryPath: P.when(O.isNone) }, (_) => {
+      return 'E_CANNOT_RESOLVE_PATH' as const
+    })
+    .with({ binaryPath: P.when(E.isRight) }, (engineInfo) => {
+      return engineInfo.binaryPath.right
+    })
+    .with({ binaryPath: P.when(E.isLeft) }, (engineInfo) => {
+      errors.push(engineInfo.binaryPath.left)
+      return 'E_CANNOT_RESOLVE_PATH' as const
+    })
+    .exhaustive()
+
+  return [`${version} (at ${path.relative(process.cwd(), absolutePath)}${resolved})`, errors] as const
+}
+
+export function resolveEngine(binaryName: BinaryType): T.Task<EngineInfo> {
+  const envVar = engineEnvVarMap[binaryName]
+  const pathFromEnv = process.env[envVar]
+
+  if (pathFromEnv && fs.existsSync(pathFromEnv)) {
+    /**
+     * Extract EngineInfo from a library engine
+     */
+    const rest = { libraryPath: O.fromNullable(pathFromEnv), fromEnvVar: O.fromNullable(envVar) }
+    const engineInfoLibraryTask = pipe(
+      safeGetBinaryVersion(pathFromEnv, binaryName),
+      TE.matchW(
+        (versionError) => ({ version: E.left(versionError), ...rest }),
+        (version) => ({ version: E.right(version), ...rest }),
+      ),
+    )
+    return engineInfoLibraryTask
+  }
+
+  /**
+   * Extract EngineInfo from a binary engine
+   */
+  const engineInfoBinaryTask: T.Task<EngineInfoBinary> = pipe(
+    safeResolveBinary(binaryName),
+    TE.matchEW(
+      (binaryPathError) => {
+        const result = T.of({ binaryPath: E.left(binaryPathError) })
+        return result as T.Task<EngineInfoBinaryPathError>
+      },
+      (binaryPath) => {
+        const rest = { binaryPath: E.right(binaryPath) }
+        const result = pipe(
+          safeGetBinaryVersion(pathFromEnv, binaryName),
+          TE.matchW(
+            (versionError) => ({ version: E.left(versionError), ...rest }),
+            (version) => ({ version: E.right(version), ...rest }),
+          ),
+        )
+        return result as T.Task<EngineInfoBinaryPathSuccess>
+      },
+    ),
+  )
+  return engineInfoBinaryTask
+}
```

Comment:
Using `fp-ts` here, makes this code cryptic, it's hard to review and maintain. (the one letter "things" don't help and it's not obvious what `matchW` vs `matchEW` actually is without looking for documentation).

Is there a way to rewrite these cryptic parts, so we don't introduce too much complexity here?

Reply:
This is a "wide" pattern matching of what's inside `safeResolveBinary`, which is a lazy promise that cannot throw any error at runtime. It acts as a destructor, doing a similar work as what `reduce` does for lists.
The `W` in `matchEW` and `matchW` stands for `Widen`: https://gcanti.github.io/fp-ts/guides/code-conventions.html#what-a-w-suffix-means-eg-chainw-or-chaineitherkw. This is needed to have proper type inference when a destructor returns two different types from its code paths (usually one for the "error" branch, another one for the "success" branch), and these types are part of the same union type.

The `E` in `matchEW` stands for `Effect`: https://gcanti.github.io/fp-ts/guides/code-conventions.html#what-an-e-suffix-means-eg-matche.

I'll see to improve the readability of this, though